### PR TITLE
Shell autocreate shell instance

### LINKED ISSUE #10480: Creating shell with current SHELL_DEFINE() is not practical
I have been porting multiple network related shells to use new shell and noticed some issues how the shells are defined.
Currently one needs to do something like this in order to define a shell.
```
SHELL_CMD_REGISTER(net, &net_commands, "Networking commands", NULL);
SHELL_UART_DEFINE(shell_transport_uart);
SHELL_DEFINE(net_shell, "net:~$ ", &shell_transport_uart, 10, SHELL_FLAG_OLF_CRLF);
```

Currently this code is called by `subsys/net/ip/net_shell.c` when `CONFIG_NET_SHELL=y` is set.

There are other network related shells that can co-exists with this network shell. For example if one defines `CONFIG_NET_L2_IEEE802154_SHELL=y` then IEEE 802.15.4 specific shell is created. It is also created similar way as generic network shell.

```
SHELL_CMD_REGISTER(ieee802154, &ieee802154_commands, "IEEE 802.15.4 commands", NULL);
SHELL_UART_DEFINE(ieee802154_shell_transport_uart);
SHELL_DEFINE(ieee802154_shell, "ieee802154:~$ ", &ieee802154_shell_transport_uart, 10, SHELL_FLAG_OLF_CRLF);
```

It really looks like that the backend configuration should not be done like this for a shell. In these examples, the actual shell implementation is in proper place but how this shell is used should be placed in some shell specific directory and controlled by generic config options. The shells in question do not really care about the backend, they are just generic shells that can be used with any backend.

It looks like that the `SHELL_DEFINE()` could easily just have the two parameters like this `SHELL_DEFINE(net_shell, SHELL_FLAG_OLF_CRLF);`. The other parameters should be defined in some other file.

Also noticed that if I enable various shells at the same time, and this is quite common case with networking, the history of the shell stops working.

### PR DESCRIPTION
**Pull request by jakub-uC**
With this PR user will not have to bother creating shell backend. All that is required from the user is to enable particular shell backend in the Kconfig file. For each enabled backend shell instance will be created automatically.

Next step is to add console backend to the shell. Once this will be in place it will be enough to activate shell commands to start default shell instance: console.

Fixes #10480

### CODE REVIEW

File: tests/bluetooth/shell/src/main.c [hunk 1]
```diff
@@ -26,17 +26,13 @@
 
 #include <gatt/hrs.h>
 
-#define print(_sh, _ft, ...) \
+#define print_sh(_sh, _ft, ...) \
```
Comment (carlescufi) [inline tests/bluetooth/shell/src/main.c]: WHy is this rename needed as part of this PR?

Comment (jarz-nordic) [reply tests/bluetooth/shell/src/main.c]: I was affraid that `print` and `printk` are too similar.

**Review by jukkar — APPROVED** [review]
Thanks for this support. I tested this with multiple networking related shells and things work now as expected.

File: tests/bluetooth/shell/src/main.c (hunk 1 above)
Comment (carlescufi) [reply tests/bluetooth/shell/src/main.c]: I don't think that's an issue. Can you remove that change?

File: tests/bluetooth/shell/src/main.c [hunk 2]
```diff
@@ -104,11 +100,9 @@ SHELL_CMD_REGISTER(hrs, &hrs_cmds, "Heart Rate Service shell commands",
 
 void main(void)
 {
-	(void)shell_init(&uart_shell, NULL, true, true, LOG_LEVEL_INF);
-
-	print(&uart_shell, "Type \"help\" for supported commands.");
-	print(&uart_shell, "Before any Bluetooth commands you must `bt init`"
-	      " to initialize the stack.");
+	printk("Type \"help\" for supported commands.");
```
Comment (carlescufi) [inline tests/bluetooth/shell/src/main.c]: Why did you change this from `print` to `printk`?

Comment (nordic-krch) [reply tests/bluetooth/shell/src/main.c]: shell_fprintf must be used only in the scope of the command.

Comment (jarz-nordic) [reply tests/bluetooth/shell/src/main.c]: `print` is a macro, in fact it is a function `shell_fprintf`. This function **shall not** be used outside of command context what is a case here. I was trying to print this welcome message by Logger.

Comment (jarz-nordic) [reply tests/bluetooth/shell/src/main.c]: What is more you do not have access to uart_shell any more because it is created in shell_uart.c file.

File: tests/bluetooth/shell/src/main.c (hunk 1 above)
Comment (carlescufi) [reply tests/bluetooth/shell/src/main.c]: Well yes, because it's unrelated to this particular problem you are solving with the PR here

File: tests/bluetooth/shell/src/main.c (hunk 2 above)
Comment (carlescufi) [reply tests/bluetooth/shell/src/main.c]: OK, that makes sense.

**Review by carlescufi — APPROVED** [review]
